aten/src/ATen/core/NestedIntSymNodeImpl.h

```diff
@@ -183,4 +184,28 @@ class TORCH_API NestedIntSymNodeImpl : public SymNodeImpl {
   int64_t coeff_;
 };
 
+class TORCH_API UnionFind {
+    std::unordered_map<int64_t, std::pair<int64_t, size_t>> data;
+public:
+    int64_t find(int64_t x) {
+        if(data.count(x) == 0) data[x] = {x, 1};
+        if (data[x].first != x) {
+            data[x].first = find(data[x].first);
+        }
+        return data[x].first;
```

Comment:
nit: This code is repeatedly looking up `data[x]` in the hash map, but I believe written carefully it can do the lookup only once. You can use `try_emplace` to insert `{x, 1}` into the map only if it does not exist, and this returns an iterator which you can use to modify the value of the map (by editing `it->second`).